service/history/nDCHistoryReplicator.go

```diff
@@ -840,33 +844,103 @@ func (r *nDCHistoryReplicatorImpl) backfillHistory(
 		runID,
 		lastEventID,
 		lastEventVersion))
+	localHistoryIterator := collection.NewPagingIterator(r.getHistoryFromLocalPaginationFn(
+		ctx,
+		branchToken,
+		lastEventID,
+	))
 
 	var lastHistoryBatch *commonpb.DataBlob
 	prevTxnID := common.EmptyVersion
-	for historyIterator.HasNext() {
-		historyBlob, err := historyIterator.Next()
+	historyBranch, err := serialization.HistoryBranchFromBlob(branchToken, enumspb.ENCODING_TYPE_PROTO3.String())
+	if err != nil {
+		return nil, err
+	}
+	latestBranchID := historyBranch.GetBranchId()
+	var prevBranchID string
+
+	sortedAncestors := copyAncestors(historyBranch.GetAncestors())
+	sortedAncestorsIdx := 0
+	historyBranch.Ancestors = nil
+
+BackfillLoop:
+	for remotehistoryIterator.HasNext() {
+		historyBlob, err := remotehistoryIterator.Next()
+		if err != nil {
+			return nil, err
+		}
+
+	HistoryExistenceCheckLoop:
+		for localHistoryIterator.HasNext() {
```

Comment:
no strong opinion but I think we can move this check loop outside the backfill loop and make the code easier to understand. 

All we need here is the last eventID in the branch and then backfill from that point. 

Another idea: Do you know if read reverse history can be used here to get the last batch directly?

Reply:
Yes. I just need last event id.